mongo/src/main/java/com/redhat/lightblue/mongo/crud/BasicDocDeleter.java

```diff
@@ -35,14 +47,103 @@
-        LOGGER.debug("Removal complete, write result={}", result);
-        response.setNumDeleted(result.getN());
+
+        int numDeleted = 0;
+
+        try (DBCursor cursor = collection.find(mongoQuery, null)) {
+            List<DocInfo> docsToDelete = new ArrayList<>();
+
+            while (cursor.hasNext()) {
+
+                if (docsToDelete.size() < batchSize) {
+                    // build batch
+                    DBObject doc = cursor.next();
+                    DocInfo docInfo = new DocInfo(doc, ctx);
+
+                    docsToDelete.add(docInfo);
+
+                    ctx.getFactory().getInterceptors().callInterceptors(InterceptPoint.PRE_CRUD_DELETE_DOC, ctx, docInfo.docCtx);
+                }
+
+                if (docsToDelete.size() == batchSize || !cursor.hasNext()) {
+                    // batch built or run out of documents
+
+                    // TODO: write concern
+                    BulkWriteOperation bw = collection.initializeUnorderedBulkOperation();
+
+                    List<Object> _ids = docsToDelete.stream().map(d -> d._id).collect(Collectors.toList());
+                    bw.find(new BasicDBObject("_id", new BasicDBObject("$in", _ids))).remove();
+
+                    docsToDelete.forEach(d -> d.docCtx.setCRUDOperationPerformed(CRUDOperation.DELETE));
+
+                    try {
+                        LOGGER.debug("Bulk deleting docs");
+                        BulkWriteResult result = bw.execute();
+                        LOGGER.debug("Bulk deleted docs - attempted {}, deleted {}", _ids.size(), result.getRemovedCount());
+                        numDeleted+=result.getRemovedCount();
+                    } catch (BulkWriteException bwe) {
+                        LOGGER.error("Bulk write exception", bwe);
+                        // TODO: will partially successful deletes work? Seems there is no BulkWriteResult if exception happens.
+                        handleBulkWriteError(bwe.getWriteErrors(), docsToDelete);
+                    } catch (RuntimeException e) {
+                        LOGGER.error("Exception", e);
+                        throw e;
+                    } finally {
+                        for (DocInfo doc : docsToDelete) {
+                            if (!doc.docCtx.hasErrors()) {
+                                ctx.getFactory().getInterceptors().callInterceptors(InterceptPoint.POST_CRUD_INSERT_DOC, ctx, doc.docCtx);
+                            }
+                        }
+                    }
+
+                    // clear list before processing next batch
+                    docsToDelete.clear();
+                }
+            }
+        }
+
+        response.setNumDeleted(numDeleted);
+    }
+
+    private void handleBulkWriteError(List<BulkWriteError> errors, List<DocInfo> docs) {
+        for (BulkWriteError e : errors) {
+            DocInfo doc = docs.get(e.getIndex());
+            // TODO: what error should we use here?
```

Comment:
Please comment on this, I'm not sure which error code to use here. Maybe we need a new one?